feat: add `WithEnv` method for setting a task's environment variables

mahadzaryab1 opened the pull request:
Added a method `WithEnv` that takes in a string array (`[]string `) and overrides the environment executing a `exec.Command` to that array. This PR fixes #80.

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1769,57 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
 	}
 }
 
+func TestWithEnv_SetEmptyEnvironment(t *testing.T) {
```
Could we be a bit more descriptive here?

```suggestion
func TestWithEnv_UnsetsAllEnvVarsGivenEmptySlice(t *testing.T) {
```

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1769,57 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
+	buf := new(bytes.Buffer)
+	env := []string{}
+
+	_, err := script.NewPipe().WithStdout(buf).WithEnv(env).Exec("printenv").Stdout()
```
If you're testing the output of some command, wouldn't it be slightly easier to call `String` on the pipe, rather than creating a buffer and using `WithStdout`?

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1769,57 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
+
+	if len(buf.String()) != 0 {
+		got := strings.Split(strings.TrimRight(buf.String(), "\n"), "\n")
+		t.Errorf("expected 0 environment variables, got %d %v", len(got), got)
```
I'm not sure the string-munging is all that helpful here, and clutters up the test a bit, making it hard to see what's going on. The point you're making is that the string should be empty. If it's not, we want to know what's in it. That's all.

```suggestion
		if output != "" {
		        t.Errorf("want empty environment, got %q", output)
		 }
```

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1769,57 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
+	buf := new(bytes.Buffer)
+	env := []string{"ENV1=test1", "ENV2=test2"}
+
+	_, err := script.NewPipe().WithStdout(buf).WithEnv(env).Exec("printenv").Stdout()
```
It seems a shame to add an implicit dependency on the `printenv` command, doesn't it? Could we use the shell instead?

mahadzaryab1 replied:
@bitfield would you please be able to elaborate on this for me? i'm not sure what it means to use the shell instead

bitfield replied:
Well, you'll notice that `script` itself doesn't rely on _any_ external commands being present—which is a good thing, because we basically can't rely on that! And the tests don't use many external commands either: currently just `go` (which seems at least reasonable) and `echo` (which exists most places, is usually a shell builtin, and even works on Windows, though with slightly different syntax).

If we could stick to those, without adding a new dependency that we don't absolutely need, that would be a nice bonus.

mahadzaryab1 replied:
@bitfield there is the `env` command but i'm not sure that's too different from using `printenv`. the other option would be to use `echo $ENV1` to capture the value of a single environment variable but i don't believe there's a way to print all the variables using `echo`. let me know what you think!

bitfield replied:
If we could avoid adding a new external dependency, that would be great!

mahadzaryab1 replied:
@bitfield using `echo` doesn't actually work since the variables aren't exported. I'm not sure if there's another way to print the variables we want without using `env` or `printenv`.

bitfield replied:
Yes, this is a bit of a subtle one. If you did something like this:

```go
Exec("echo ENV1=$ENV1")
```

that wouldn't work, because `$`-expansion is done by the _shell_, and we're not executing a shell! We're executing the `echo` command directly, which doesn't interpolate variables.

On the other hand, if we exec a shell instead:

```go
Exec("sh -c 'echo ENV1=$ENV1'")
```

this works as expected, because the shell expands `$ENV1` before passing it to `echo`.

We do use `sh -c` in a couple of existing tests, so there's no problem with depending on it here.

mahadzaryab1 replied:
That's really cool! Thanks for the explanation @bitfield

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1769,57 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	got := strings.Split(strings.TrimRight(buf.String(), "\n"), "\n")
```
I think it's probably a bad idea to munge the output in this way, and it means we're not comparing against the _actual_ output of the command, we're testing against some arbitrarily-munged version. For tests to be clear and direct, it helps to specify the exact, literal string you expect to get.

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1769,57 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
+	}
+
+	// not setting the environment should simply use the task's default environment
+	expected := os.Environ()
```
Just a house style issue, but we universally use the names `want` and `got` in tests, and the failure message "want %q, got %q" (for example).

bitfield reviewed (changes requested):
Great work so far! Just a couple of points:

* Do these tests work on Windows? (You can add Windows-specific tests to `script_windows_test.go`)
* What potential bugs in the environment-setting code are we failing to catch in tests here?

mahadzaryab1 commented:
> Great work so far! Just a couple of points:
> 
> * Do these tests work on Windows? (You can add Windows-specific tests to `script_windows_test.go`)
> * What potential bugs in the environment-setting code are we failing to catch in tests here?

Thank you so much for the great feedback @bitfield! I addressed some of your feedback and had a couple of follow-up questions: 
- Is it possible to run the windows-specific tests locally? (I'm on my MacBook)
- For potential bugs - one possibility could be passing in malformed environment variables. Should we add some tests for this? Let me know if there's something else that I'm missing.

bitfield commented:
> * Is it possible to run the windows-specific tests locally? (I'm on a MacBook)

That's a good question—I'd like to be able to do this too. I can run them in CI, but of course it would be helpful for you to get the results locally. What about running the tests in a Windows Docker image? Would that work?

> * For potential bugs - one possibility could be passing in malformed environment variables. Should we add some tests for this? Let me know if there's something else that I'm missing.

Well, that's really the user's problem! We don't do anything but pass on what they gave us to the `exec.Cmd`, which is absolutely right. The kind of bugs I meant are the ones in *our* logic. Basically, we could forget to set `p.Env`, and we could forget to reference it in either `p.Exec` or `p.ExecForEach`, and I think the existing tests cover those. So we're good.

There is still the awkward question of what happens if someone writes some arbitrary `Filter` function on the pipe that does something like `os.Getenv`. You could imagine users thinking that `p.WithEnv` should affect this, but it won't. As far as I can see there's no way around this, unless _we_ called `os.Setenv`, which is exactly what we don't want to do. Maybe we could clarify this in the documentation somehow?

mahadzaryab1 commented:
@bitfield do you know if the windows tests run in WSL?

bitfield commented:
> @bitfield do you know if the windows tests run in WSL?

I don't.

lambrospetrou commented on script.go:
```diff
@@ -898,6 +911,14 @@ func (p *Pipe) WithStdout(w io.Writer) *Pipe {
 	return p
 }
 
+// WithEnv sets the pipe's environment to the string array env. This will override
```
One small suggestion, which relates to the discussion on the main page too, is that this doesn't modify the "pipe" environment, but the "next exec command's" environment. If thie documentation mentions this I believe it's clear and won't be confusing to folks accessing the env outside of the `Exec` functions.

mahadzaryab1 replied:
@lambrospetrou thanks for the suggestion! i clarified this comment now to add more detail about the usage. let me know what you think and if you have any other feedback

bitfield commented on script.go:
```diff
@@ -34,6 +34,12 @@ type Pipe struct {
+	// Each entry in the array should be of the form key=value.
+	// If env is not nil, it will replace the default environment variables
+	// when executing commands.
+	env []string
```
Is this protected by the mutex? Does it need to be?

mahadzaryab1 replied:
this should be protected by the mutex based on the race condition that https://github.com/bitfield/script/pull/209 fixes! i'll make the change

bitfield commented on script.go:
```diff
@@ -898,6 +911,15 @@ func (p *Pipe) WithStdout(w io.Writer) *Pipe {
 	return p
 }
 
+// WithEnv sets the environment for the exec commands to the string array env.
```
```suggestion
// WithEnv sets the environment for subsequent [Pipe.Exec] and [Pipe.ExecForEach] commands to the string slice env.
```

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1768,52 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
+	}
+}
+
+func TestWithEnv_SetMultipleEnvVars(t *testing.T) {
```
Can we make this a sentence describing the behaviour?

bitfield commented on script.go:
```diff
@@ -34,6 +34,12 @@ type Pipe struct {
+	// env contains the environment to run any exec commands with.
+	// Each entry in the array should be of the form key=value.
+	// If env is not nil, it will replace the default environment variables
+	// when executing commands.
```
I think it's worth pointing out also that if `env` is an _empty_ array, it will remove everything, and commands will be run with a completely empty environment. Since `nil` and empty arrays usually have the same semantics, this would otherwise be slightly surprising (but completely correct) behaviour.

bitfield replied:
The place to say this, of course, would be in the doc comment for `WithEnv`—in fact, we don't really need to say anything much about the struct field itself.

bitfield commented on script.go:
```diff
@@ -858,6 +871,15 @@ func (p *Pipe) Wait() {
+// WithEnv sets the environment for subsequent [Pipe.Exec] and [Pipe.ExecForEach] commands
+// to the string slice env. This will override the default process environment variables
+// when executing commands run via [Pipe.Exec] or [Pipe.ExecForEach]. This will not affect
+// the environment outside of [Pipe.Exec] or [Pipe.ExecForEach].
```
"The environment outside" is a bit vague—shall we say explicitly that it doesn't affect the current _process's_ environment, in the way that `os.Setenv` would? (I can imagine people getting confused that this doesn't behave the way `Setenv` does.)

bitfield commented on script.go:
```diff
@@ -858,6 +871,15 @@ func (p *Pipe) Wait() {
+// when executing commands run via [Pipe.Exec] or [Pipe.ExecForEach]. This will not affect
+// the environment outside of [Pipe.Exec] or [Pipe.ExecForEach].
+func (p *Pipe) WithEnv(env []string) *Pipe {
+	p.env = env
```
We also need the mutex here.

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1768,52 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
+	}
+}
+
+func TestWithEnvChangesExecutionEnvironment(t *testing.T) {
```
```suggestion
func TestWithEnv_SetsGivenVariablesForSubsequentExec(t *testing.T) {
```

In case you didn't know, which by the way would be absolutely excusable, the `_` is a hint to [gotestdox](https://github.com/bitfield/gotestdox). It's a tool for printing out your test names as descriptive sentences (try it in this repo!), but without the `_` it doesn't know that `WithEnv` is a camel-cased function name and not just the words "with env".

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1768,52 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
+	}
+}
+
+func TestNotSettingEnvFallsBackToDefaultEnvironment(t *testing.T) {
```
I don't honestly think we need this test (and that's not something I often find myself saying!)

After all, it's essentially saying that if you don't use `WithEnv`, nothing special happens—and that was already the case.

bitfield reviewed (changes requested):
Nice work so far!

mahadzaryab1 commented:
@bitfield addressed your feedback! What should we do about the `TestWithEnv_UnsetsAllEnvVarsGivenEmptySlice` test case? It currently relies on the printenv command and is failing on Windows.

bitfield commented:
> @bitfield addressed your feedback! What should we do about the `TestWithEnv_UnsetsAllEnvVarsGivenEmptySlice` test case? It currently relies on the printenv command and is failing on Windows. We can't echo a specific environment variable in this case since we're not setting anything.

Hmm, what about echoing some variable which would usually be there (`HOME` or `PATH`, for example), and checking that it's empty? That proves that `WithEnv` erased the environment.

We'll need a different one on Windows (`HOMEPATH`)?

mahadzaryab1 commented:
> // env contains the environment to run any exec commands with.
> 	// Each entry in the array should be of the form key=value.
> 	// If env is not nil, it will replace the default environment variables
> 	// when executing commands.

@bitfield I tried a different approach where I do an `os.SetEnv` and test that it doesn't show up. when passing in an empty slice. What do you think of this approach? This was mostly to keep things platform-agnostic so we don't have to write different tests for different platforms.

bitfield commented on script_test.go:
```diff
@@ -1768,6 +1768,44 @@ func TestWithStdout_SetsSpecifiedWriterAsStdout(t *testing.T) {
 
+func TestWithEnv_UnsetsAllEnvVarsGivenEmptySlice(t *testing.T) {
+	t.Parallel()
+	os.Setenv("ENV1", "test1")
```
Not a bad idea in principle, but unfortunately `Setenv` isn't safe to use in a parallel test, because it affects the environment for all tests. (If we _were_ going to do that, we could use `t.Setenv`, which does the cleanup automatically, but it's still not parallelisable.)

Actually, what we care about is not the whole _test binary's_ environment, but just the _pipe's_, isn't it? When you frame it that way, all we need to do is create a pipe with some environment variable already set, and check that `WithEnv` unsets it. Something like this, perhaps?

```go
func TestWithEnv_UnsetsAllEnvVarsGivenEmptySlice(t *testing.T) {
	t.Parallel()
	p := script.NewPipe().WithEnv([]string{"ENV1=test1"}).Exec("sh -c 'echo ENV1=$ENV1'")
	want := "ENV1=test1\n"
	got, err := p.String()
	if err != nil {
		t.Fatal(err)
	}
	if got != want {
		t.Fatalf("want %q, got %q", want, got)
	}
	got, err = p.Echo("").WithEnv([]string{}).Exec("sh -c 'echo ENV1=$ENV1'").String()
	if err != nil {
		t.Fatal(err)
	}
	want = "ENV1=\n"
	if got != want {
		t.Errorf("want %q, got %q", want, got)
	}
}
```

mahadzaryab1 replied:
Great idea! Thanks @bitfield

bitfield reviewed (approved):
Great job, @mahadzaryab1, thanks a lot! Everybody wants to _propose_ features, and even do quick-and-dirty implementations of them, but that's easy. Few people are willing to do the slow, detailed, careful, high-quality implementation and polish that characterises durable software. You've done that on several issues, and that makes you `script`'s most valuable contributor by far!

Anything else you'd like to work on? I think we have enough to put a release together now, but if you could pick an existing issue to solve, or even suggest a new one, what would it be?

mahadzaryab1 commented:
> Great job, @mahadzaryab1, thanks a lot! Everybody wants to _propose_ features, and even do quick-and-dirty implementations of them, but that's easy. Few people are willing to do the slow, detailed, careful, high-quality implementation and polish that characterises durable software. You've done that on several issues, and that makes you `script`'s most valuable contributor by far!
> 
> Anything else you'd like to work on? I think we have enough to put a release together now, but if you could pick an existing issue to solve, or even suggest a new one, what would it be?

@bitfield Thank you so much for your kind words and guidance in helping me working on these issues. I love writing Go and learnt a lot about the language from your code reviews! In terms of proposing new features, I would love to and I'll definitely start to give this more thought and open issues for any ideas that I have. However, in the meantime, I also do want to keep contributing to `script` so I wanted to ask if there were any features needed implementing? I would love to contribute in whatever way that I can and whatever it is that the users of this package require.

bitfield commented:
Great! Well, have a look at the open issues—there are quite a few of them, and most are waiting for some sort of design or further consideration, so I'm sure your input would be very helpful here!
